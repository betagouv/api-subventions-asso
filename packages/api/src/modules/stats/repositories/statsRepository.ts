import { RoleEnum } from "../../../@enums/Roles";
import { DefaultObject } from "../../../@types";
import db from "../../../shared/MongoConnection";
import { englishMonthNames } from "../../../shared/helpers/DateHelper";
import { NbRequestsPerMonthRequest } from "@api-subventions-asso/dto";

export class StatsRepository {
    private readonly collection = db.collection("log");

    public async countUsersByRequestNbOnPeriod(
        start: Date,
        end: Date,
        nbReq: number,
        includesAdmin: boolean
    ): Promise<number> {
        const matchQuery: { $match: DefaultObject } = {
            $match: {
                timestamp: {
                    $gte: start,
                    $lte: end
                }
            }
        };
        if (!includesAdmin) {
            matchQuery.$match["meta.req.user.roles"] = { $nin: [RoleEnum.admin] };
        }

        return (
            (
                await this.collection
                    .aggregate([
                        matchQuery,
                        { $group: { _id: "$meta.req.user.email", nbOfRequest: { $sum: 1 } } },
                        { $match: { nbOfRequest: { $gte: nbReq } } },
                        { $count: "nbOfUsers" }
                    ])
                    .next()
            )?.nbOfUsers || 0
        ); // If no stats found nbOfUsers is null but whant retrun an number
    }

    public async countMedianRequestsOnPeriod(start: Date, end: Date, includesAdmin: boolean): Promise<number> {
        const buildQuery = () => {
            const matchQuery: { $match: DefaultObject } = {
                $match: {
                    timestamp: {
                        $gte: start,
                        $lte: end
                    },
                    "meta.req.user.email": { $ne: null }
                }
            };
            if (!includesAdmin) {
                matchQuery.$match["meta.req.user.roles"] = { $nin: [RoleEnum.admin] };
            }

            return [
                matchQuery,
                { $group: { _id: "$meta.req.user.email", nbOfRequest: { $sum: 1 } } },
                { $sort: { nbOfRequest: 1 } }
            ];
        };

        const result = await this.collection.aggregate(buildQuery()).toArray();

        if (!result.length) return 0;

        const middle = Math.floor(result.length / 2);

        if (result.length % 2 === 0) {
            return (result[middle - 1].nbOfRequest + result[middle].nbOfRequest) / 2;
        }

        return result[middle].nbOfRequest;
    }

    public async countRequestsPerMonthByYear(year: number, includesAdmin: boolean): Promise<NbRequestsPerMonthRequest> {
        const start = new Date(year, 0, 1);
        const end = new Date(year + 1, 0, 0);
        const buildQuery = () => {
            const matchQuery: { $match: DefaultObject } = {
                $match: {
                    timestamp: {
                        $gte: start,
                        $lte: end
                    },
                    "meta.req.user.email": { $ne: null }
                }
            };
            if (!includesAdmin) {
                matchQuery.$match["meta.req.user.roles"] = { $nin: [RoleEnum.admin] };
            }
            const annotateQuery = { $addFields: { month: { $month: "$timestamp" } } };

            return [matchQuery, annotateQuery, { $group: { _id: "$month", nbOfRequest: { $count: {} } } }];
        };

        const queryResult = await this.collection.aggregate(buildQuery()).toArray();
        const resultByMonth0Index = {};
        for (const { _id, nbOfRequests } of queryResult) {
            resultByMonth0Index[_id - 1] = nbOfRequests;
        }
        return englishMonthNames.reduce((acc, month, index) => {
            acc[month] = resultByMonth0Index[index] || 0;
            return acc;
        }, {});
    }
}

const statsRepository = new StatsRepository();
export default statsRepository;
