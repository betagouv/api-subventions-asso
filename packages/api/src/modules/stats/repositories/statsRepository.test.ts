import statsRepository from "./statsRepository";

const FIRST_DAY_YEAR = new Date(Date.UTC(2022, 0, 1));
const YEAR = 2022;

const dateFactory = (diff: number) => {
    const date = new Date(FIRST_DAY_YEAR);
    date.setDate(FIRST_DAY_YEAR.getDate() + diff);
    return date;
};

describe("StatsRepository", () => {
    describe("countUsersByRequestNbOnPeriod()", () => {
        const NB_REQUESTS = 3;
        const START = dateFactory(-1);
        const END = dateFactory(1);

        it("should call mongo with admin filters", async () => {
            const mock = jest
                // @ts-expect-error statsRepository.collection is private attribute
                .spyOn(statsRepository.collection, "aggregate")
                // @ts-expect-error type error due to mocking documents
                .mockImplementationOnce(() => ({ next: () => ({}) }));

            await statsRepository.countUsersByRequestNbOnPeriod(START, END, NB_REQUESTS, false);

            expect(mock.mock.calls).toMatchSnapshot();
        });

        it("should call mongo without admin filters", async () => {
            const mock = jest
                // @ts-expect-error statsRepository.collection is private attribute
                .spyOn(statsRepository.collection, "aggregate")
                // @ts-expect-error type error due to mocking documents
                .mockImplementationOnce(() => ({ next: () => ({}) }));

            await statsRepository.countUsersByRequestNbOnPeriod(START, END, NB_REQUESTS, true);

            expect(mock.mock.calls).toMatchSnapshot();
        });
    });

    describe("countMedianRequestsOnPeriod()", () => {
        const START = dateFactory(-1);
        const END = dateFactory(1);

        it("should call mongo with admin filters", async () => {
            const mock = jest
                // @ts-expect-error statsRepository.collection is private attribute
                .spyOn(statsRepository.collection, "aggregate")
                // @ts-expect-error type error due to mocking documents
                .mockImplementationOnce(() => ({ toArray: () => [] }));

            await statsRepository.countMedianRequestsOnPeriod(START, END, false);

            expect(mock.mock.calls).toMatchSnapshot();
        });

        it("should call mongo without admin filters", async () => {
            const mock = jest
                // @ts-expect-error statsRepository.collection is private attribute
                .spyOn(statsRepository.collection, "aggregate")
                // @ts-expect-error type error due to mocking documents
                .mockImplementationOnce(() => ({ toArray: () => [] }));

            await statsRepository.countMedianRequestsOnPeriod(START, END, true);

            expect(mock.mock.calls).toMatchSnapshot();
        });

        it("should return median (odd)", async () => {
            // @ts-expect-error statsRepository.collection is private attribute
            jest.spyOn(statsRepository.collection, "aggregate").mockImplementationOnce(() => ({
                // @ts-expect-error type error due to mocking documents
                toArray: () => [{ nbOfRequest: 1 }, { nbOfRequest: 2 }, { nbOfRequest: 3 }]
            }));
            const expected = 2;
            const actual = await statsRepository.countMedianRequestsOnPeriod(START, END, true);

            expect(actual).toBe(expected);
        });

        it("should return median (pair)", async () => {
            // @ts-expect-error statsRepository.collection is private attribute
            jest.spyOn(statsRepository.collection, "aggregate").mockImplementationOnce(() => ({
                // @ts-expect-error type error due to mocking documents
                toArray: () => [{ nbOfRequest: 1 }, { nbOfRequest: 2 }, { nbOfRequest: 3 }, { nbOfRequest: 4 }]
            }));
            const expected = 2.5;
            const actual = await statsRepository.countMedianRequestsOnPeriod(START, END, true);

            expect(actual).toBe(expected);
        });
    });

    describe("countRequestsPerMonthByYear()", () => {
        async function checkMonthlyAvgStatFormat(aggregateOutput, expected) {
            // @ts-expect-error statsRepository.collection is private attribute and mock typing errors
            jest.spyOn(statsRepository.collection, "aggregate").mockReturnValueOnce({ toArray: () => aggregateOutput });
            const actual = await statsRepository.countRequestsPerMonthByYear(YEAR, false);
            expect(actual).toStrictEqual(expected);
        }

        it("should format the result properly", async () => {
            await checkMonthlyAvgStatFormat(
                [
                    { _id: 1, nbOfRequests: 201 },
                    { _id: 2, nbOfRequests: 21 },
                    { _id: 3, nbOfRequests: 20 },
                    { _id: 4, nbOfRequests: 201 },
                    { _id: 5, nbOfRequests: 13 },
                    { _id: 6, nbOfRequests: 201 },
                    { _id: 7, nbOfRequests: 201 },
                    { _id: 8, nbOfRequests: 15 },
                    { _id: 9, nbOfRequests: 201 },
                    { _id: 10, nbOfRequests: 300 },
                    { _id: 11, nbOfRequests: 201 },
                    { _id: 12, nbOfRequests: 1 }
                ],
                {
                    January: 201,
                    February: 21,
                    March: 20,
                    April: 201,
                    May: 13,
                    June: 201,
                    July: 201,
                    August: 15,
                    September: 201,
                    October: 300,
                    November: 201,
                    December: 1
                }
            );
        });

        it("should fill in empty months", async () => {
            await checkMonthlyAvgStatFormat(
                [
                    { _id: 1, nbOfRequests: 201 },
                    { _id: 2, nbOfRequests: 21 },
                    { _id: 10, nbOfRequests: 300 },
                    { _id: 12, nbOfRequests: 1 }
                ],
                {
                    January: 201,
                    February: 21,
                    March: 0,
                    April: 0,
                    May: 0,
                    June: 0,
                    July: 0,
                    August: 0,
                    September: 0,
                    October: 300,
                    November: 0,
                    December: 1
                }
            );
        });
    });
});
