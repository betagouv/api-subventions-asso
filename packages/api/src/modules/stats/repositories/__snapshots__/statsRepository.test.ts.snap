// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsRepository countUsersByRequestNbOnPeriod() should call mongo with admin filters 1`] = `
Array [
  Array [
    Array [
      Object {
        "$match": Object {
          "meta.req.user.roles": Object {
            "$nin": Array [
              "admin",
            ],
          },
          "timestamp": Object {
            "$gte": 2021-12-31T00:00:00.000Z,
            "$lte": 2022-01-02T00:00:00.000Z,
          },
        },
      },
      Object {
        "$group": Object {
          "_id": "$meta.req.user.email",
          "nbOfRequest": Object {
            "$sum": 1,
          },
        },
      },
      Object {
        "$match": Object {
          "nbOfRequest": Object {
            "$gte": 3,
          },
        },
      },
      Object {
        "$count": "nbOfUsers",
      },
    ],
  ],
]
`;

exports[`StatsRepository countUsersByRequestNbOnPeriod() should call mongo without admin filters 1`] = `
Array [
  Array [
    Array [
      Object {
        "$match": Object {
          "timestamp": Object {
            "$gte": 2021-12-31T00:00:00.000Z,
            "$lte": 2022-01-02T00:00:00.000Z,
          },
        },
      },
      Object {
        "$group": Object {
          "_id": "$meta.req.user.email",
          "nbOfRequest": Object {
            "$sum": 1,
          },
        },
      },
      Object {
        "$match": Object {
          "nbOfRequest": Object {
            "$gte": 3,
          },
        },
      },
      Object {
        "$count": "nbOfUsers",
      },
    ],
  ],
]
`;
